server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java

```diff
@@ -968,35 +1067,38 @@ public CompletableFuture<Revision> destroyToken(Author author, String appId) {
         requireNonNull(author, "author");
         requireNonNull(appId, "appId");
 
-        final String commitSummary = "Delete the token: " + appId;
+        final String commitSummary = "Destroy the token: " + appId;
         final UserAndTimestamp userAndTimestamp = UserAndTimestamp.of(author);
 
-        final ContentTransformer<JsonNode> transformer = new ContentTransformer<>(
-                TOKEN_JSON, EntryType.JSON, node -> {
-            final Tokens tokens = tokens(node);
+        final TokensTransformer transformer = new TokensTransformer((headRevision, tokens) -> {
             final Token token = tokens.get(appId); // Raise an exception if not found.
             if (token.deletion() != null) {
-                throw new IllegalArgumentException("The token is already deleted: " + appId);
+                throw new ChangeConflictException("The token is already destroyed: " + appId);
             }
 
-            final ImmutableMap.Builder<String, Token> appIdsBuilder = ImmutableMap.builder();
-            for (Entry<String, Token> entry : tokens.appIds().entrySet()) {
-                if (!entry.getKey().equals(appId)) {
-                    appIdsBuilder.put(entry);
-                } else {
-                    final String secret = token.secret();
-                    assert secret != null;
-                    appIdsBuilder.put(appId, new Token(token.appId(), secret, token.isSystemAdmin(),
-                                                       token.isSystemAdmin(), token.creation(),
-                                                       token.deactivation(), userAndTimestamp));
-                }
-            }
-            final Map<String, Token> newAppIds = appIdsBuilder.build();
-            return Jackson.valueToTree(new Tokens(newAppIds, tokens.secrets()));
+            final String secret = token.secret();
+            assert secret != null;
+            final Token newToken = new Token(token.appId(), secret, token.isSystemAdmin(),
+                                             token.isSystemAdmin(), token.creation(),
+                                             token.deactivation(), userAndTimestamp);
+            return new Tokens(newAppIds(tokens, appId, newToken), tokens.secrets());
         });
         return tokenRepo.push(INTERNAL_PROJECT_DOGMA, Project.REPO_DOGMA, author, commitSummary, transformer);
     }
 
+    private static Map<String, Token> newAppIds(Tokens tokens, String appId, Token newToken) {
+        final ImmutableMap.Builder<String, Token> appIdsBuilder =
+                ImmutableMap.builderWithExpectedSize(tokens.appIds().size());
+        for (Entry<String, Token> entry : tokens.appIds().entrySet()) {
+            if (!entry.getKey().equals(appId)) {
+                appIdsBuilder.put(entry);
+            } else {
+                appIdsBuilder.put(appId, newToken);
+            }
+        }
```

Comment:
Side note: It seems useful if we have an `ImmutableMap` utility that helps to update or remove.
```java
class ImmutableMapUtil {
    Map<K, V> remove(Map<K, V> old, K key) { 
      // create a new map and remove `key`.
      ... 
    }

    Map<K, V> add(Map<K, V> old, K key, V value) { ... }
}
```

Reply:
That looks useful. 👍 Let me try that later. 😉